x/payment/types/wirepayfordata.go

```diff
@@ -118,9 +130,36 @@ func (msg *MsgWirePayForData) ValidateBasic() error {
 		}
 	}
 
+	if err := msg.ValidateAllSquareSizesCommitedTo(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ValidateAllSquareSizesCommitedTo returns an error if the list of square sizes
+// committed to don't match all squares sizes expected for this message size.
+func (msg *MsgWirePayForData) ValidateAllSquareSizesCommitedTo() error {
+	allSquareSizes := AllSquareSizes(int(msg.MessageSize))
+	sort.Slice(allSquareSizes, func(i, j int) bool { return allSquareSizes[i] < allSquareSizes[j] })
+
+	committedSquareSizes := msg.committedSquareSizes()
+	sort.Slice(committedSquareSizes, func(i, j int) bool { return committedSquareSizes[i] < committedSquareSizes[j] })
+
+	if !reflect.DeepEqual(allSquareSizes, committedSquareSizes) {
```

Comment:
[for reviewers] we don't need to use `reflect.DeepEqual` here but I used it for simplicity. If performance is critical on this code path then we can consider using a custom Equals function:

Ref: https://yourbasic.org/golang/compare-slices/

Reply:
given the simplicity of using something else, imo I think we should. While the performance gained or lost is arguably trivial compared to checking that all of the commitments are actually accurate, I still think we should scavenge as much performance as possible given https://github.com/informalsystems/audit-celestia/issues/6

Also, even a small percentage will add up over time since this function will be called by every full node for every message submitted.